Skill Trees: Skeleton Code

jackandthebean opened the pull request:
# Skill Trees: Skeleton Code
### Following up on initial implementation in Sprint 2
We devoted the first two sprints to the design of the `x/rpg-skilltrees` module. Towards the end of the second sprint, @VannStrasen began the initial write-up of our implementation (#527), drawing from the data structures and function headers we developed during our team design discussions. He closed #527 with the following:
> I initially wrote this code in a branch called `skilltrees`, which I realized was less descriptive than necessary, so I shifted the code over to `skilltrees/skeleton_code`. Because I also did not use the correct format for `.c` and `.h` files (they were in the wrong folder), the `Makefile` was not formatted properly. While it was compiling correctly on my `Makefile`, I am not sure if it will compile correctly in the chiventure `Makefile`, because I haven't been able to work on that yet. As Sprint 2 is ending now, I'm not issuing a pull request until the Makefile is finished—that'll be work for the beginning of the next sprint.

Now I have added a properly formatted `CMake` file (#561), updated the top-level chiventure `CMake` file, and our skeleton code builds a static library without error, so I am creating this pull request to incorporate the foundational skill trees code into the `dev` branch.
### Specific files to be added
Specifically, our skeleton code consists of two files: `skilltrees.h` and `skilltrees.c`, which implement the structs that will hold chiventure-supported skills, as well as their associated new, init, and free functions. The skeleton code also implements basic functions that manipulate skill lists—adding/removing skills—and retrieve skill types that already exist within a given skill list. We are currently developing the more advanced skill tree functions and look forward to incorporating those as well, following the approval of this foundational pull request.

firatciftci commented on include/skilltrees/skilltrees.h:
```diff
@@ -0,0 +1,90 @@
+    int (*execute_skill)(int param1, int param2);
+};
+
+typedef struct skill_t skill_t;
```
Minor comment on style: You can have this written in a simpler manner:
```c
typedef struct skill {
.
.
.
} skill_t;
```

firatciftci commented on include/skilltrees/skilltrees.h:
```diff
@@ -0,0 +1,90 @@
+// Struct which holds a player's every skill, differentiated by passive and
+// active.
+// When not holding a skill, each space in the passive or active skill lists is
+// defined as NULL.
```
Comment on style: You should use multi-line comments (`/* ... */`) instead of regular comments (`// ...`) for function headers. So this would look like the following:
```c
/* struct_name
 *
 * active_skills: [description]
 * cur_num_active: [description]
 * ...
 */
```

The same goes for function header comments, where instead of struct field descriptions, you should list the input parameters and also the return value.

firatciftci commented on include/skilltrees/skilltrees.h:
```diff
@@ -0,0 +1,90 @@
+    unsigned int max_num_passive;
+};
+
+typedef struct allskills_t allskills_t;
```
The name `allskills_t` seems a bit vague to me; without looking at the source code, it's easy to mistake it with a struct for all of the skills in the game. To make it a bit clearer, I think a name such as `player_skills_t` would be better.

firatciftci commented on include/skilltrees/skilltrees.h:
```diff
@@ -0,0 +1,90 @@
+    unsigned int experience;
+    char* description;
+    int (*execute_skill)(int param1, int param2);
+};
```
You might want to add an integer field `id` to keep track of the skills, instead of relying on the `name` field. This way, you would be able to have two skills with the same name (assuming they have different IDs). This would also let you create new `skill_t` structs with just an integer value, rather than dealing with string manipulation from the get-go.

firatciftci commented on include/skilltrees/skilltrees.h:
```diff
@@ -0,0 +1,90 @@
+// Creates a new skill list, initially mallocing space for 5 active and 5
+// passive skills.
+// Returns the malloc'd list.
+allskills_t* create_skill_list();
```
You might want to take in the max space for active and passive skills as function parameters, instead of hardcoding it to 5 for each by default.

jackandthebean commented:
@firatciftci thank you for the feedback! I made all the changes you suggested, reworked all the code to match the style of all our assignments more closely, increased the flexibility of the skill "effect" function pointers, and added a `skill_execute()` function which simply executes the function pointed to by a given `skill_t` struct.

jackandthebean commented:
I will put my closing statement for this closed PR at the end of #561.

firatciftci commented:
Issue Score: ✔️++

Comments:
Great work on this PR!

<!-- {"score": "check-plus-plus"} -->
